src/Nixfmt/Predoc.hs

```diff
@@ -264,33 +265,36 @@ isComment _ = False
 
 --- Manually force a group to its compact layout, by replacing all relevant whitespace.
 --- Does not recurse into inner groups.
-unexpandSpacing :: Doc -> Doc
-unexpandSpacing [] = []
-unexpandSpacing ((Spacing Space) : xs) = Spacing Hardspace : unexpandSpacing xs
-unexpandSpacing ((Spacing Softspace) : xs) = Spacing Hardspace : unexpandSpacing xs
-unexpandSpacing ((Spacing Break) : xs) = unexpandSpacing xs
-unexpandSpacing ((Spacing Softbreak) : xs) = unexpandSpacing xs
-unexpandSpacing (s@(Spacing _) : xs) = s : unexpandSpacing xs
-unexpandSpacing (x : xs) = x : unexpandSpacing xs
+spanEnd :: (a -> Bool) -> Seq a -> (Seq a, Seq a)
+spanEnd p = fmap Seq.reverse . Seq.spanl p . Seq.reverse
 
-spanEnd :: (a -> Bool) -> [a] -> ([a], [a])
-spanEnd p = fmap reverse . span p . reverse
+unexpandSpacing :: Doc -> Doc
+unexpandSpacing = foldMap \case
+  Spacing Space -> [Spacing Hardspace]
+  Spacing Softspace -> [Spacing Hardspace]
+  Spacing Break -> []
+  Spacing Softbreak -> []
+  x -> [x]
 
 -- Manually force a group to its compact layout, by replacing all relevant whitespace.
 -- Does recurse into inner groups.
 -- An optional maximum line length limit may be specified.
 -- Fails if the doc contains hardlines or exceeds the length limit
 unexpandSpacing' :: Maybe Int -> Doc -> Maybe Doc
-unexpandSpacing' (Just n) _ | n < 0 = Nothing
-unexpandSpacing' _ [] = Just []
-unexpandSpacing' n (txt@(Text _ _ _ t) : xs) = (txt :) <$> unexpandSpacing' (n <&> subtract (textWidth t)) xs
-unexpandSpacing' n (Spacing Hardspace : xs) = (Spacing Hardspace :) <$> unexpandSpacing' (n <&> subtract 1) xs
-unexpandSpacing' n (Spacing Space : xs) = (Spacing Hardspace :) <$> unexpandSpacing' (n <&> subtract 1) xs
-unexpandSpacing' n (Spacing Softspace : xs) = (Spacing Hardspace :) <$> unexpandSpacing' (n <&> subtract 1) xs
-unexpandSpacing' n (Spacing Break : xs) = unexpandSpacing' n xs
-unexpandSpacing' n (Spacing Softbreak : xs) = unexpandSpacing' n xs
-unexpandSpacing' _ (Spacing _ : _) = Nothing
-unexpandSpacing' n ((Group _ xs) : ys) = unexpandSpacing' n $ xs <> ys
+unexpandSpacing' _ Seq.Empty = Just []
+unexpandSpacing' mn (x :<| xs)
+  | Just n <- mn, n < 0 = Nothing
+  | otherwise =
+      let unexpandSubtract t = unexpandSpacing' (subtract t <$> mn)
+      in case x of
+          txt@(Text _ _ _ t) -> (txt :<|) <$> unexpandSubtract (textWidth t) xs
+          (Spacing Hardspace) -> (Spacing Hardspace :<|) <$> unexpandSubtract 1 xs
+          (Spacing Space) -> (Spacing Hardspace :<|) <$> unexpandSubtract 1 xs
+          (Spacing Softspace) -> (Spacing Hardspace :<|) <$> unexpandSubtract 1 xs
+          (Spacing Break) -> unexpandSpacing' mn xs
+          (Spacing Softbreak) -> unexpandSpacing' mn xs
+          (Spacing _) -> Nothing
+          ((Group _ ws)) -> unexpandSpacing' mn $ ws <> xs
```

Comment:
This also changes the semantics a bit, because `unexpandSpacing' (Just -1) Seq.Empty` now returns `Just []` instead of `Nothing`. This is likely the cause behind the changes in https://github.com/infinixbot/nixpkgs/commit/2fb6f2c261a13bc1f72a7d2c8a338b57853bce95. Even if that would be better, let's have this PR only be internal refactorings.

Reply:
Sure, this was an accident, nice catch!